### ISSUE TITLE
Premature Contrail Death In Met With Supersaturated RHi

### ISSUE THREAD

**Issue by Calebsakhtar** [contrailcirrus/pycontrails#241]
**Description**
I have hijacked pycontrails to run over custom met data as opposed to ERA5 data, and I have observed premature contrail death in cases where the relative humidity was enforced to be supersaturated throughout the met domain.

![fig_alt_vs_t](https://github.com/user-attachments/assets/4c4b58b7-22e5-48ef-a85f-587def914162)
Figure 1: Contrail centre altitude vs time since formation.

![fig_RHi_vs_t](https://github.com/user-attachments/assets/ad645908-e655-463d-8a1c-91f780eda5bc)
Figure 2: Contrail centre RHi vs time since formation.

![fig_I_vs_t](https://github.com/user-attachments/assets/8c0bb85f-c5a8-4330-a627-20caa0a2a5ff)
Figure 3: Contrail ice mass per unit depth vs time since formation.

Observations:
Figure 3 shows a large reduction in the ice mass for the 115% and 120% RHi cases. In theory, the contrail should grow in ice mass indefinitely (as long as it remains in supersaturated air) until an end-of-life condition is reached. However, this reduction in ice mass happens much sooner than the contrail death for both cases above.

The remaining cases should also continue to grow in ice mass. 

**Details**
- Version: 0.1.dev1781
- Module: [e.g. units, datalib.ecmwf]
- OS: Windows

**Steps to Reproduce**
The steps and code for reproduction are all available in the following repo: https://github.com/Calebsakhtar/pycontrails-debugging

**Comment by Calebsakhtar** [comment]
I am very aware that my code might be what is causing the unexpected behavior - so I would love to hear from one of the pycontrails developers.

Apologies for the disturbance @mlshapiro, @thabbott, but I am not sure who to ping for this. I kindly request that either one of you points me towards someone that could help with this.

**Comment by thabbott** [comment]
Tagging @zebengberg too.

@Calebsakhtar: thanks for opening the issue--we'll figure out who to assign to this and one of us will take a look soon!

**Comment by Calebsakhtar** [comment]
Hello @thabbott, just bumping this.

**Comment by thabbott** [comment]
Hi @Calebsakhtar--sorry for taking so long to get to this!

A few quick notes:
- the `rad` object that you pickled still refers to a file on your machine, so I had to replace it with my own data to be able to run your MWE.
- Cocip limits contrail depth to 1500 m, which is affecting your results with higher RHi. You can disable it by adding `"max_depth": np.inf` to your Cocip parameters.

The equation that Cocip uses to evolve ice mass is d(M(I + qs))/dt = qs RHi dM/dt, where M is the contrail mass, I is the ice mass concentration, and qs is saturation specific humidity. Your plot of "total ice mass" is essentially M I (the segment length doesn't vary much). dM/dt > 0 (which should be true as long as the contrail is in supersaturated air) means M(I + qs) has to increase, but M I doesn't necessarily have to. In your simulations, qs increases as the contrail sinks. I think you'd see a monotonic increase if you plotted M(I + qs) instead--have you tried that?

So, I don't think what you're seeing is a bug in the sense that the code is doing something different than the Cocip equations say it should. (Though maybe I will have to retract that if your plot of M(I + qs) doesn't look the way I think it should.) There's still the question of whether the Cocip equations make physical sense, though. I think there could be an issue in the way it interacts with sedimentation... let me think a little bit more about this.

**Comment by Calebsakhtar** [comment]
Hi @thabbott.

I have re evaluated the MWE with the changes you suggested. Here is a plot of the total water content in the contrail (I + qsat)*M.

![fig_H20_vs_t](https://github.com/user-attachments/assets/993b4308-fa45-4991-a459-8a6c3034a568)

This makes sense according to your explanatioj, but I still fail to understand why the ice content of the contrail decreases given that the ambient air is supersaturated and not depleting. This behaviour appears to go against what I think would make sense thermodynamically: there is water available above supersaturation, and there are plenty of nucleation spots. 

I will think about it further, but I am not entirely convinced that this behaviour is physical yet.

I'll get back to you soon.

**Comment by thabbott** [comment]
I can think of one case where this behavior is physical and one where it's not.

If qs is increasing because the contrail is embedded in subsiding air, then this behavior seems reasonable. Subsidence won't change the total water (ice + vapor) in the contrail, but it can change the partitioning between ice and vapor. Cocip assumes that the air within the contrail is at equilibrium, so subsidence will cause ice to sublimate by increasing qs.

In your simulations, though, I think the contrail altitude is probably increasing because of sedimentation. (Do you know whether w is non-zero in your met data?) Ice particles falling through a layer where RHi = 1 will always be at equilibrium with the surrounding air, so there should be no sublimation or deposition. I'm not sure that Cocip gets this right, though... 

Looking at the [code](https://github.com/contrailcirrus/pycontrails/blob/aa3256c109b860bd386690b993ad461af9fac421/pycontrails/models/cocip/cocip.py#L2222) it seems like Cocip saves initial values of I and qs, updates a bunch of plume properties (including altitude, which changes due to vertical advection *and* sedimentation), then computes a final value for I based on the initial and final values of qs. This means that changes in qs will have the same effect on I regardless whether they're caused by subsidence or sedimentation, and I think these two cases actually need to be handled separately.

**Comment by thabbott** [comment]
(A more succinct way of describing what I think is going wrong is that the equation that Cocip uses to evolve ice mass is derived assuming that ice particles and water vapor follow a common trajectory, and that assumption doesn't hold when there's sedimentation.)

**Comment by thabbott** [comment]
How hard would it be to set up a test case with w = 0, RHi = 1.2 above 9900 m, and RHi = 1 below? It seems obvious to me that M I should be constant once the contrail falls into the region where RHi = 1. (Do you agree?) I'm guessing Cocip will instead produce a decrease in M I, though, and I'd regard that as clear evidence that there's a problem with the way Cocip's updating ice mass.

**Comment by thabbott** [comment]
> @thabbott In a zero-shear case I agree that such a test case should produce a constant M. In a case with nonzero shear, it is my understanding that M will continue to grow as the contrail shears.

Sorry, my plaintext math is probably hard to parse. I agree that M might grow, but I think M*I should be constant.

> Making such a test case will not take much effort - you will hopefully hear back from me on the weekend with the results.

Ok, great!

> However, I would have thought that in a model like CoCiP where the ice particles move with the air together this should not matter. From the point of view of the particles, the conditions should remain constant.

Conditions typically aren't constant in a subsiding parcel--pressure and usually temperature will increase, which affects saturation specific humidity, which changes the partitioning between ice and water vapor in a parcel with a fixed amount of total water.

Maybe an important conceptual point: ice particles in Cocip aren't directly in contact with air with the ambient relative humidity from the met data. Cocip models a contrail as a plume that contains a mixture of ice and water vapor at equilibrium, so RHi in the plume interior is always 1. Ambient RHi only influences plume properties through mixing between the plume and ambient air.

**Comment by Calebsakhtar** [comment]
Hi @thabbott I had a go at the test case you requested (see the latest commit of the mwe repo above):

Altitude vs RHi
![fig_tristan_alt_vs_rhi](https://github.com/user-attachments/assets/18ea15c8-1d19-49eb-8099-2549a4564373)

M * ( I + qs) vs t
![fig_tristan_H20_vs_t](https://github.com/user-attachments/assets/5f34a098-ab8c-4859-ad03-2bd8d340f736)

M * I
![fig_tristan_I_vs_t](https://github.com/user-attachments/assets/65ce5293-2b51-4b76-9197-15219427abaa)

> It seems obvious to me that M I should be constant once the contrail falls into the region where RHi = 1. (Do you agree?)
I agree, but as you predicted, CoCiP decreases M*I. I have to agree with you about this being unwanted behaviour.

> Conditions typically aren't constant in a subsiding parcel--pressure and usually temperature will increase, which affects saturation specific humidity, which changes the partitioning between ice and water vapor in a parcel with a fixed amount of total water.
Thank you for that, I will take this into accont for future endeavours.

> Maybe an important conceptual point: ice particles in Cocip aren't directly in contact with air with the ambient relative humidity from the met data. Cocip models a contrail as a plume that contains a mixture of ice and water vapor at equilibrium, so RHi in the plume interior is always 1. Ambient RHi only influences plume properties through mixing between the plume and ambient air.
I was unaware of this and suddenly some of the behaviour we are seeing makes sense. Basically it's as if CoCiP skips the settling process and reaches the steady-state solution for the distribution of water at each timestep (given the ambient conditions). In practice, nucleation and sublimation takes time. I suspect that the relative humidity would only approach ~100% in the centre of the contrail, where the effects from mixing are minimal.

> Sorry, my plaintext math is probably hard to parse.
Apologies, it was late and I was tired! (and the letters I and M got split into new lines)

**Comment by thabbott** [comment]
@Calebsakhtar awesome, this is a great test case!

I worked through a change (just on paper so far) that I think will fix this issue. I'll try it in a branch of your repo so I can use the test case you've put together and let you know whether it works the way I think it will.

> Basically it's as if CoCiP skips the settling process and reaches the steady-state solution for the distribution of water at each timestep (given the ambient conditions).

Yes, exactly!

> In practice, nucleation and sublimation takes time. I suspect that the relative humidity would only approach ~100% in the centre of the contrail, where the effects from mixing are minimal.

I don't think it would be straightforward to modify CoCiP so that RHi varies in space within the plume, but there's no reason you couldn't replace the current saturation adjustment with a finite phase relaxation time. My guess is that the reason CoCiP uses saturation adjustment is that early CoCiP simulations often used a time step of 1 hour, which is long compared to the phase relaxation time at least in young contrails. Now that most people run CoCiP with a shorter time step it might be worth revisiting this assumption.

**Comment by Calebsakhtar** [comment]
Hi @thabbott. 


> I worked through a change (just on paper so far) that I think will fix this issue. I'll try it in a branch of your repo so I can use the test case you've put together and let you know whether it works the way I think it will.

Sounds great! I am looking forward to your implementation.

I had a chat about this with Seb and other academics in the field, and they helped me understand why the contrail was dying.
In particular, the contrail saturated air falls with the ice crystals in CoCiP and heats up due to the atmospheric lapse rate. This heating, by definition, makes the air less-than saturated. Because CoCiP enforces RHi = 1 if ice is present, this means that the ice mass can decrease if the rate of uptake of water is less than the rate of decrease of available water above saturation. I think I finally got there!

Now that I finally understand the picture, I can suggest a potential fix. To achieve more physical results, the amount of water in the contrail should be artificially updated to ensure no ice particles are lost as long as the ambient air is at or above saturation. I have heard that this has been done in other models before so I am not claiming any originality.


> Now that most people run CoCiP with a shorter time step it might be worth revisiting this assumption.

From what I have seen, I don't think it's an unreasonable assumption as long as the ice crystals are not too large and the timestep is not too small, but I agree that looking into this would improve insight. In particular, some graphs of the "average RHi" vs timestep from LES studies would be nice.

**Comment by thabbott** [comment]
> In particular, the contrail saturated air falls with the ice crystals in CoCiP and heats up due to the atmospheric lapse rate. This heating, by definition, makes the air less-than saturated. Because CoCiP enforces RHi = 1 if ice is present, this means that the ice mass can decrease if the rate of uptake of water is less than the rate of decrease of available water above saturation. I think I finally got there!

Yep, that's my understanding of what's going wrong, too. (Just to confirm: the people you talked to agree that this is unphysical, right? The surrounding water vapor shouldn't actually fall alongside the ice crystals.)

> Now that I finally understand the picture, I can suggest a potential fix. To achieve more physical results, the amount of water in the contrail should be artificially updated to ensure no ice particles are lost as long as the ambient air is at or above saturation. I have heard that this has been done in other models before so I am not claiming any originality.

Yeah, that's basically what my fix is, though I don't think the additional source of water is artificial. If you work through a water budget for the volume occupied by a contrail's ice crystals there's a 100% real water vapor source associated with sedimentation into a lower layer with higher specific humidity.

**Comment by thabbott** [comment]
I took a stab at a fix and am getting this for the nice test case you put together:
![fig_tristan_RHi_vs_t](https://github.com/user-attachments/assets/21978af0-393f-4966-97dc-9c74da0cde39)
![fig_tristan_I_vs_t](https://github.com/user-attachments/assets/ca956112-2bcd-4e62-b081-51988617e9f7)

This looks pretty good to me. The changes I made should conserve M*I to numerical precision if RHi is exactly 100%, and I think the slow upward drift is because interpolation can produce RHi values a bit above 100% even when RHi is exactly 100% in met data (see https://py.contrails.org/notebooks/specific-humidity-interpolation.html).

I opened a PR in your pycontrails repo if you want to see the changes. You probably don't want to merge the PR (I deleted some workflow files because they were causing issues for me), but it should be easy to copy the changes if you want to test them yourself.

**Comment by thabbott** [comment]
One other note: you can reduce errors from humidity interpolation and control the upward drift a bit better if you add `"interpolation_q_method": "log-q-log-p" to your Cocip parameters:

![fig_tristan_I_vs_t](https://github.com/user-attachments/assets/2591b134-6251-4530-82fc-498cd5aba71d)

**Comment by Calebsakhtar** [comment]
> (Just to confirm: the people you talked to agree that this is unphysical, right? The surrounding water vapor shouldn't actually fall alongside the ice crystals.)

Yes, they did!


> If you work through a water budget for the volume occupied by a contrail's ice crystals there's a 100% real water vapor source associated with sedimentation into a lower layer with higher specific humidity.

Agreed, makes sense to me!


> This looks pretty good to me. 

Agreed, looks very promising. Thanks for doing this! I will be doing the validation using my other test cases as well. In theory the contrails in supersaturated air should keep growing.


> The changes I made should conserve M*I to numerical precision if RHi is exactly 100%, and I think the slow upward drift is because interpolation can produce RHi values a bit above 100% even when RHi is exactly 100% in met data (see https://py.contrails.org/notebooks/specific-humidity-interpolation.html).

This is fair enough. It also doesn't help that water budget requires the use of saturation humidities, whose functions of temperature vary slightly depending on the definitions. That plus the interpolation issues you mentioned (from the fact that q_sat(T) increases with positive curvature) means that ice growth is almost inevitable at 100% RHi cases. Fortunately, such a case with perfectly uniform ambient RHi at saturation is aphysical so I think this is acceptable.


> One other note: you can reduce errors from humidity interpolation and control the upward drift a bit better if you add `"interpolation_q_method": "log-q-log-p" to your Cocip parameters:

I will keep this in mind, thank you kindly.

@thabbott It looks like you have done a great job with this. Thank you so much for the time and energy spent on this issue.

Just a final few questions for you: what sort of validation would be required for this to go into the latest dev version of pycontrails? What is a reasonable estimate for when this could be merged with the pycontrails dev branch? How can I help expedite the process?
